Training related changes to the onnxruntime website (#15900)
microsoft/onnxruntime: docs/get-started/on-device-training.md

---
title: On-Device Training
parent: Get Started
nav_order: 12
---

# On-Device Training with ONNX Runtime
{: .no_toc }
`On-Device Training` refers to the process of training a model on an edge device, such as mobile phones, embedded devices, gaming consoles, web browsers, etc. This is in contrast to training a model on a server or a cloud. Training on the device:

- Is also useful for the task of personalization where the model needs to be trained on the user's data.
- Can be used for tasks such as federated learning where the model is locally trained on data that is distributed across multiple devices in an effort to build a more robust aggregated global model.
- Can improve data privacy and security, especially when working with sensitive data that cannot be shared with a server or a cloud.

> - Can also be beneficial in scenarios where network connectivity is unreliable or limited.

Reviewer: Not convinced

> `ONNX Runtime Training` offers an easy way to efficiently train and infer a wide range of ONNX models on edge devices. The training process is divided into two phases:

Reviewer: "a wide range" -> can be removed

Author: I kind of like having `a wide range` to show that we support many models.
We are not quantifying the number of models, but the hope is that any model that can be onboarded with inference can also be used with on-device training.

But can drop it if you feel strongly about this.

> ---

Reviewer 2: for the file name, is that better we call it training-on-device.md?

Author: I think `on-device-training.md` might be better considering this is the marketing name we want to go with.

Any reason to use `training-on-device` as the file name?

Reviewer 2: It's fine :). I mean, this will make training-on-device.md and training-pytorch.md stay close when since the filenames are listed in alphabet order.

Author: Updated the name of the file. :)

title: On-Device Training
parent: Get Started
nav_order: 12
---

# On-Device Training with ONNX Runtime
{: .no_toc }
`On-Device Training` refers to the process of training a model on an edge device, such as mobile phones, embedded devices, gaming consoles, web browsers, etc. This is in contrast to training a model on a server or a cloud. Training on the device

> 

Reviewer 3: colon after Training on the device.

- Can improve data privacy and security, especially when working with sensitive data that cannot be shared with a server or a cloud.

> - It is also useful for the task of personalization where the model needs to be trained on the user's data.

Reviewer 3: Should be "Is useful..." Also the ordering should be personalization first and privacy second.

> - It can also be beneficial in scenarios where network connectivity is unreliable or limited.

Reviewer 3: Should be "Can also be..."

Separately should also add the use case of improving a global model with user data through Federated Learning.

> In addition to the above scenarios, there could be many more that warrant the need for `On-Device Training`.

Reviewer 3: Can remove this line.

`ONNX Runtime Training` offers an easy way to efficiently train and infer a wide range of ONNX models on edge devices. The training process is divided into two phases:
- [the offline phase](#the-offline-phase)
- [the training phase](#the-training-phase).

> ## The Offline Phase

Reviewer 3: Good to call out that this phase does not use user data on the cloud.

Reviewer: Shall we call this Training Artifact Generation Phase? In my experience, "offline phase" is a bit confusing to some people.

> In this phase, training artifacts are prepared on a server, cloud or a desktop. These artifacts can be generated by using the `onnxruntime-training`'s [artifact generation python tools](./../api/python/learning_on_the_edge/training_artifacts.html) available in the python package.

Reviewer 2: ```suggestion
In this phase, training artifacts are prepared on a server, cloud or a desktop. These artifacts can be generated by using the `onnxruntime-training`'s [artifact generation Python tools](./../api/python/learning_on_the_edge/training_artifacts.html) available in the Python package.
```

Refer to the [installation instructions](./../install/index.md#offline-phase---prepare-for-training)

## The Training Phase
Once these artifacts are generated, they can be deployed to production scenarios on edge devices. `ONNX Runtime` offers a wide range of packages in multiple language bindings.

Refer to the [installation instructions](./../install/index.md#training-phase---on-device-training) for a complete list of all language bindings.

> Once training on the edge device is complete, an inference ready onnx model can be generated on the edge device itself. This model can then be used for inferencing with onnxruntime.

Reviewer 2: ```suggestion
Once training on the edge device is complete, an inference-ready ONNX model can be generated on the edge device itself. This model can then be used for ONNX Runtime inferencing .
```

[comment]: <> (Learn more from the blogs. Links to the blogs go here.)

## Installation

Refer to the [installation instructions](./../install/index.md#install-for-on-device-training) for details on how to install for your scenario.

## Building from Source

Refer to the [build instructions](./../build/training.md#build-for-on-device-training) for details on how to build for your custom scenario.

[comment]: <> (Learn more from the tutorials. Links to the demo and website tutorial go here.)
[comment]: <> (Also link to the onnxruntime-training-examples repo goes here.)

> 

Reviewer 3: Here create a placeholder for link to step by step tutorial.

Author: I think having an empty placeholder might not be very meaningful right now. It may confuse the user.

I added a comment so I can address this in a subsequent PR.

## Feature Request, Bug Report or Help Needed

> In case you need help, please open an [issue](https://github.com/microsoft/onnxruntime/issues/new?assignees=&labels=training&projects=&template=06-training.yml&title=%5BTraining%5D+). We will be sure to respond to your request.

Reviewer 2: ```suggestion
In case you need help, please open an [issue](https://github.com/microsoft/onnxruntime/issues/new?assignees=&labels=training&projects=&template=06-training.yml&title=%5BTraining%5D+). 
```

do we want to write our commitment on the doc? :)
